### PR TITLE
Denon MC7000: Handle autoloop button in script

### PR DESCRIPTION
**Pull request by flosse**
I was confused by the Auto-Loop button because I can't toggle the state like expected, so I solved the issue by copying the implementation of `Pioneer-DDJ-SX-scripts.js`.

### CODE REVIEW

**Comment by Be-ing** [comment]
Please rebase this on the 2.3 branch.

**Comment by Swiftb0y** [comment]
@toszlanyi what do you think?

**Comment by flosse** [comment]
> Please rebase this on the 2.3 branch.

sure! Just force pushed the branch.

File: res/controllers/Denon-MC7000-scripts.js [hunk 1]
```diff
@@ -837,6 +837,19 @@ MC7000.censor = function(channel, control, value, status, group) {
     }
 };
 
+// Auto-Loop Button
+MC7000.autoLoop = function(channel, control, value, status, group) {
+    if (value) {
+        if (engine.getValue(group, "loop_enabled")) {
+            engine.setValue(group, "reloop_toggle", true);
+            engine.setValue(group, "reloop_toggle", false);
```
Comment (Swiftb0y) [inline res/controllers/Denon-MC7000-scripts.js:845]:
```suggestion
            script.toggleControl(group, "reloop_toggle");
```
Same below

**Review by Swiftb0y — CHANGES_REQUESTED** [review]

File: res/controllers/Denon-MC7000-scripts.js (hunk 1 above)
Comment (flosse) [reply res/controllers/Denon-MC7000-scripts.js]: cool, thanks!

**Comment by flosse** [comment]
BTW:
I'm going to remove the code style issues in a separate commit soon.

**Comment by Swiftb0y** [comment]
We usually squash-merge mapping changes anyways.

**Comment by flosse** [comment]
I don't understand the script linting... first it tells me to replace `var` with `const` and now it's mucking around with `The keyword 'const' is reserved` :-1:

**Comment by Be-ing** [comment]
> I don't understand the script linting... first it tells me to replace var with const and now it's mucking around with The keyword 'const' is reserved -1

:/ please ignore that one. That's eslint being overzealous. The JavaScript interpreter in 2.3 doesn't support `const`, though that will be fixed in Mixxx 2.4.

**Comment by flosse** [comment]
> :/ please ignore that one.

Ok, I see.. I'm sorry for the massive github action runs but I can't cancel them and the only chance for me was to see what the linter tells me on each iteration.

**Comment by Swiftb0y** [comment]
> I don't understand the script linting... first it tells me to replace var with const and now it's mucking around with The keyword 'const' is reserved

I'm sorry, this is a temporary issue. We have upgraded the javascript environment in 2.4 and with that, new linter rules apply as well. So when you base a branch on main (so 2.4) the ES7 linter rules will get applied (suggesting to replace var with const). However, that makes the script incompatible with 2.3. So make sure you are basing your work on the correct branch.

**Comment by flosse** [comment]
> I'm sorry, this is a temporary issue.

no problem, i was just confused ;-)

> So make sure you are basing your work on the correct branch.

OK :+1: so then I'll try to make it work with 2.3.

Besides of this I just wonder if the new behavior suits for most users or should I also introduce a boolean flag to keep the old behavior?

**Comment by Swiftb0y** [comment]
> Besides of this I just wonder if the new behavior suits for most users or should I also introduce a boolean flag to keep the old behavior?

I don't think thats necessary. But I'd like to hear @toszlanyi's opinion on this

File: res/controllers/Denon-MC7000-scripts.js [hunk 2]
```diff
@@ -837,15 +837,24 @@ MC7000.censor = function(channel, control, value, status, group) {
...
 
+// Auto-Loop Button
+MC7000.autoLoop = function(channel, control, value, status, group) {
+    if (engine.getValue(group, "loop_enabled")) {
+        script.toggleControl(group, "reloop_toggle");
+    } else {
+        script.toggleControl(group, "beatloop_activate");
+    }
+};
+
```
Comment (toszlanyi) [inline res/controllers/Denon-MC7000-scripts.js:848]: Looks reasonable to me. Just wanna check what happens if a loop is set but inactive. Does it trigger a new loop at play position or reloop the existing one... Can do that by Thursday

Comment (toszlanyi) [reply res/controllers/Denon-MC7000-scripts.js:848]: I like that little improvement.

as all `var` were replaced with `let` then the script doesnt work in Mixxx 2.3 any longer. I have therefore quickly changed my one and could push this to #4021 if you like. Or we let this one still move.

Comment (toszlanyi) [reply res/controllers/Denon-MC7000-scripts.js:848]: I couldn't get `beatloop_activate` to work properly if there was no loop set in the track yet. Anyhow `beatloop_4_activate` works to set a new loop and activate it. Is this a bug or what else do we miss here?

Comment (toszlanyi) [reply res/controllers/Denon-MC7000-scripts.js:848]: I found another issue ... Assumed there was a loop inside the track and I reloop it with `Shift` and `Autoloop`, go out the loop by pressing `Autoloop` then the loop position sometimes move to the current play position and activates. This is quite unpredictable - it doesn't happen always but now and then. Expecting the loop to stay where it was but not toggled on to current play position as I did only wanted to exit the `Reloop`.

Comment (uklotzde) [reply res/controllers/Denon-MC7000-scripts.js:848]: I am using a much more complex logic for this purpose:

https://github.com/mixxxdj/mixxx/blob/de1853c1caf4a4b0d891b8052d617890fb88f2e6/res/controllers/Denon-MC6000MK2-scripts.js#L948

Don't ask me why I came up with this, I don't remember.

We need a domain specific API to unify this behavior for multiple controllers. Manipulating individual control values is too low level. Needing to write a custom function `hasLoop()` only to determine if a valid loop range is defined clearly indicates this conceptual gap.

Comment (toszlanyi) [reply res/controllers/Denon-MC7000-scripts.js:848]: @uklotzde - Awesome! Thanks for the hint. To add a check for an existing loop in works to solve and set the initial loop if no one was set yet.

```
MC7000.autoLoop = function(channel, control, value, status, group) {
    if (engine.getValue(group, "loop_in") || engine.getValue(group, "loop_enabled")) {
        script.toggleControl(group, "reloop_toggle", 100);
    } else {
        script.toggleControl(group, "beatloop_activate", 100);
    }
};
```
To leave reloop without to set a new activated loop needs to be fixed yet.

Comment (Swiftb0y) [reply res/controllers/Denon-MC7000-scripts.js:848]: > script.toggleControl(group, "reloop_toggle", 100);

The third parameter (`100`) doesn't do anything.

Comment (flosse) [reply res/controllers/Denon-MC7000-scripts.js:848]: hm...it still does not what I'd expect.
I think I have to read the API docs more accurate to understand the difference behaviors between
`loop_*`, `reloop_*` and `beatloop_*`

Comment (toszlanyi) [reply res/controllers/Denon-MC7000-scripts.js]: So what would be the workflow that you expect?

Comment (flosse) [reply res/controllers/Denon-MC7000-scripts.js]: The method `script.toggleControl` does not work for me.
At least the last commit (c43ac92)  is working at least for 90% of my usecases:
Create and activate a beatloop or disable a loop.

Comment (Swiftb0y) [reply res/controllers/Denon-MC7000-scripts.js]: That commit seems to have some significant flaws in it (see eslint output). Also make sure you are using the correct Control Objects...

File: res/controllers/Denon-MC7000-scripts.js [hunk 3]
```diff
@@ -837,6 +837,20 @@ MC7000.censor = function(channel, control, value, status, group) {
...
+        if (engine.getValue(group, "loop_enabled")) {
+            engine.setValue(group, "loop_enabled", false);
+            engine.setValue(group, "reloop_toggle", true);
+        } else {
+            engine.setValue(group, "beatloop_activate", true);
+            engine.setValue(group, "beatloop_activate", false);
+        }
+    }
+};
+
```
Comment (toszlanyi) [inline res/controllers/Denon-MC7000-scripts.js:853]: This looks a bit strange to me but I cannot test it as I built up another gear right now...

Shifted buttons do exactly what is written on the controller -> `Reloop`
No need to take care about.

File: res/controllers/Denon-MC7000-scripts.js [hunk 4]
```diff
@@ -837,6 +837,20 @@ MC7000.censor = function(channel, control, value, status, group) {
...
 
+// Auto-Loop Button
+MC7000.autoLoop = function(channel, control, value, status, group) {
+    // TODO: handle shift button
+    if (value) { // Button Down == 0x7F
+        if (engine.getValue(group, "loop_enabled")) {
+            engine.setValue(group, "loop_enabled", false);
+            engine.setValue(group, "reloop_toggle", true);
+        } else {
+            engine.setValue(group, "beatloop_activate", true);
```
Comment (uklotzde) [inline res/controllers/Denon-MC7000-scripts.js:848]: This invocation sequence requires a comment, i.e. why you need a negative edge trigger. Does this even work as intended?

File: res/controllers/Denon-MC7000-scripts.js (hunk 2 above)
Comment (toszlanyi) [reply res/controllers/Denon-MC7000-scripts.js]: > > script.toggleControl(group, "reloop_toggle", 100);
> 
> The third parameter (`100`) doesn't do anything.

Mmhhhh interesting - got this line from the [Mixxx Controls page inside the manual](https://manual.mixxx.org/2.3/en/chapters/appendix/mixxx_controls.html#controlpotmeter-controls). See last example in that linked chapter.

File: res/controllers/Denon-MC7000-scripts.js (hunk 3 above)
Comment (flosse) [reply res/controllers/Denon-MC7000-scripts.js]: You're right, the shift button is working as intended.

And yes, it looks strange, but on Saturday it worked more than 6 hours without issues :smile:

Comment (toszlanyi) [reply res/controllers/Denon-MC7000-scripts.js]: > And yes, it looks strange, but on Saturday it worked more than 6 hours without issues :smile:
> 

I could try it and indeed it works great. Cheers!

File: res/controllers/Denon-MC7000-scripts.js (hunk 4 above)
Comment (uklotzde) [reply res/controllers/Denon-MC7000-scripts.js:896]: This comment is imprecise and wrong. script.toggleControl() works as expected and inverts a boolean value, i.e. 0/false -> 1/true or 1/true -> 0/false. Setting `beatloop_activate` to `true` and then to `false` is a completely different use case.

How about setting `beatloop_activate` to `true` on *button down* and to `false` otherwise (*button up* or when entering the other `if` arm on *button down*)?

I am using this pattern a lot in the MC6000MK2 script to avoid getting stuck in inconsistent states, e.g. when a button is pressed *shifted* and released *unshifted* or the other way round. All control values that might be affected by a button and all invocation sequences must be considered.

**Review by uklotzde — CHANGES_REQUESTED** [review]

File: res/controllers/Denon-MC7000-scripts.js [hunk 5]
```diff
@@ -513,9 +513,9 @@ MC7000.PadButtons = function(channel, control, value, status, group) {
             }
         }
     } else if (MC7000.PADModeCueLoop[deckNumber]) {
-        return;
+        // TODO
```
Comment (Swiftb0y) [inline res/controllers/Denon-MC7000-scripts.js:516]: Please revert these unrelated changes.

**Review by Swiftb0y — CHANGES_REQUESTED** [review]

**Comment by toszlanyi** [comment]
Suggest to close this PR as the desired function was implemented globally with #4328

**Comment by flosse** [comment]
I use CDJs now, so this PR is no longer relevant to me :stuck_out_tongue_winking_eye:
